apps/prairielearn/src/middlewares/clientFingerprint.ts

```diff
@@ -0,0 +1,48 @@
+import asyncHandler = require('express-async-handler');
+import * as sqldb from '@prairielearn/postgres';
+import { IdSchema } from '../lib/db-types';
+import { idsEqual } from '../lib/id';
+
+const sql = sqldb.loadSqlEquiv(__filename);
+
+export default asyncHandler(async (req, res, next) => {
+  if (!res.locals.assessment_instance) {
+    throw new Error('Assessment Instance is not pressent');
+  }
+
+  const user_session_id = await sqldb.queryOptionalRow(
+    sql.select_user_session_id,
+    { session_id: req.session.id },
+    IdSchema,
+  );
+
+  const params = {
+    ip_address: req.ip,
+    // We are passing the authn user id here. However, we are checking in the SQL query 'update_assessment_instance_fingerprint' that the authn user is the owner of the assessment. This will keep us from inadvertently recording a fingerprint change for an instructor viewing the assessment instance.
```

Comment:
Nit: we typically break long comments into multiple lines so that they're visible in an editor without horizontal scrolling.

Reply:
This comment has been broken up into multiple lines in commit [400c659](https://github.com/PrairieLearn/PrairieLearn/pull/8798/commits/400c659365c3a292b31a5ad64370d98314f3ff25).